### DISCUSSION TITLE
keyboard usage

### DISCUSSION THREAD

**Discussion by emdete** [fyne-io/fyne#3533] [Q&A] [answered]
i did not find any doc about using / binding the keyboard. the demo app has the advanced tab which contains code to show the key pressed but the doc seems to lack keyboard entirely - or am i wrong?

**Comment by andydotxyz** [comment, marked as answer]
Which APIs to use would depend on what you are truing to do with keyboard events.
Is this desktop specific, or should it work for soft keyboards on mobile too? Should it be about key events or getting textual input? Or are you looking to implement keyboard shortcuts?

There is a great video that @Jacalz did at last FyneConf that could help to describe the architecture and where to look for information: https://www.youtube.com/watch?v=TDae1Kmxu_0

**Reply by emdete** [reply]
i'm not a big fan of such videos but will take a look. i wonder why there are near to no example programs (not talking about the demo app which is kind of advertising by showing everything). an example would have code for just one solution of a problem/demand to understand the solution. this makes it very difficult to understand and learn fyne.

**Reply by Jacalz** [reply]
> i wonder why there are near to no example programs

A lot of documentation exists on https://developer.fyne.io. I think that for example https://developer.fyne.io/extend/numerical-entry could be a good start. There are lot of interfaces that can be implemented so I would suggest looking at that video, like Andy said.

**Reply by andydotxyz** [reply]
> i wonder why there are near to no example programs ... an example would have code for just one solution of a problem/demand to understand the solution.

The answer is because what you ask for is a massive undertaking - even if we did manage to create an app to individually demonstrate every API then it would be basically impossible for users to find.

That said, other apps that handle key interaction:
* https://github.com/andydotxyz/fynegameboy.git
* https://github.com/fyne-io/examples/blob/develop/fractal/main.go
* https://github.com/fyne-io/terminal

Like I asked before what are you specifically trying to do? There is a lot of possible stuff to show but until we know more it's not easy to provide the right example.

**Comment by emdete** [comment]
In fact i was looking for a way to end a program via ^Q or such stuff. I found `SetOnKey*` functions that seem to allow exactly that. Would this the fyne-way to go?

**Comment by andydotxyz** [comment]
If you are looking for a key combination then the `Shortcut` API is what you want.
Probably `Window.Canvas().AddShortcut()`

**Reply by emdete** [reply]
yes, that's more like it! thank you!

but it doesnt work in my case: https://codeberg.org/Belphegor/fyne_examples/src/branch/master/keybinding/main.go

can you tell, what's wrong?

**Reply by andydotxyz** [reply]
You will have to be more specific about what "doesn't work" means, sorry.

**Reply by emdete** [reply]
sorry, it just doesnt call the function behind the shortcut if i press escape while the other functions report it just fine.

**Reply by andydotxyz** [reply]
Perhaps that relates to this: https://github.com/fyne-io/fyne/issues/2455. If not please open a new issue

**Reply by emdete** [reply]
no, it's not on mac and it seems that if i set any modifier it just works. when i create the CustomShortcut i just leave out the Modifier or set if to 0, both do not work. if i set it to KeyModifierAlt or KeyModifierControl it just works. how do i set "no modifier" correctly?

**Comment by andydotxyz** [comment]
However specifically for quit your OS should handle those shortcuts - matching their standards.

**Reply by emdete** [reply]
i am lazy on testing (or learning fyne) so i love a shortcut (even like escape) to terminate my tested program. and i wanted to understand event handling a bit more.

so first of all you are right: i wouldnt do quit differently for normal programs i write.
